File: tags/pre-release-0.2/src/ConstraintSolver/Arc2D.cpp
#include "Arc2D.h"

#include "IndependentDOF.h"
#include "DependentDOF.h"

using namespace std;
using namespace GiNaC;

// create an arc
Arc2D::Arc2D (double s_center, double t_center, double theta_1, double theta_2, double radius, SketchPlanePointer sketch_plane,
			  bool s_center_free, bool t_center_free, bool theta_1_free, bool theta_2_free, bool radius_free):
s_center_(new IndependentDOF(s_center,s_center_free)),
t_center_(new IndependentDOF(t_center,t_center_free)),
theta_1_(new IndependentDOF(theta_1,theta_1_free)),
theta_2_(new IndependentDOF(theta_2,theta_2_free)),
radius_(new IndependentDOF(radius,radius_free)),
Edge2DBase(sketch_plane)
{
	dof_list_.push_back(s_center_);
	dof_list_.push_back(t_center_);
	dof_list_.push_back(theta_1_);
	dof_list_.push_back(theta_2_);
	dof_list_.push_back(radius_);

	point1_ = GeneratePoint1();
	point2_ = GeneratePoint2();
}

Arc2D::Arc2D (DOFPointer s_center, DOFPointer t_center, DOFPointer theta_1, DOFPointer theta_2, DOFPointer radius, SketchPlanePointer sketch_plane):
s_center_(s_center),
t_center_(t_center),
theta_1_(theta_1),
theta_2_(theta_2),
radius_(radius),
Edge2DBase(sketch_plane)
{
	dof_list_.push_back(s_center_);
	dof_list_.push_back(t_center_);
	dof_list_.push_back(theta_1_);
	dof_list_.push_back(theta_2_);
	dof_list_.push_back(radius_);

	point1_ = GeneratePoint1();
	point2_ = GeneratePoint2();
}

void Arc2D::Get3DLocations(double & x_center, double & y_center, double & z_center)
{
	sketch_plane_->Get3DLocation(s_center_->GetValue(), t_center_->GetValue(), x_center, y_center, z_center);
}


void Arc2D::ApplySelectionMask(SelectionMask mask)
{
	if(mask == All || mask == Edges || mask == Arcs)
		SetSelectable(true);
	else
		SetSelectable(false);
}


// Return a point that will follow the first endpoint of the arc
// This point will use dependent DOF's to define its location
Point2DPointer Arc2D::GeneratePoint1()
{
	// Create expressions defining s and t coordinates of the first endpoint of the arc
	ex s_1 = s_center_->GetVariable() + radius_->GetVariable()*cos(theta_1_->GetVariable());
	ex t_1 = t_center_->GetVariable() + radius_->GetVariable()*sin(theta_1_->GetVariable());

	// create DOF lists for each DOF
	std::vector <DOFPointer> s_1_dof_list;
	s_1_dof_list.push_back(s_center_);
	s_1_dof_list.push_back(radius_);
	s_1_dof_list.push_back(theta_1_);

	std::vector <DOFPointer> t_1_dof_list;
	t_1_dof_list.push_back(t_center_);
	t_1_dof_list.push_back(radius_);
	t_1_dof_list.push_back(theta_1_);

	// create dependent DOF's based on the above expressions
	DOFPointer s_dof(new DependentDOF(s_1, s_1_dof_list));
	DOFPointer t_dof(new DependentDOF(t_1, t_1_dof_list));

	// create the actual point object
	Point2DPointer result(new Point2D(s_dof, t_dof, sketch_plane_));
	return result;
}

// Return a point that will follow the second endpoint of the arc
// This point will use dependent DOF's to define its location
Point2DPointer Arc2D::GeneratePoint2()
{
	// Create expressions defining s and t coordinates of the second endpoint of the arc
	ex s_2 = s_center_->GetVariable() + radius_->GetVariable()*cos(theta_2_->GetVariable());
	ex t_2 = t_center_->GetVariable() + radius_->GetVariable()*sin(theta_2_->GetVariable());

	// create DOF lists for each DOF
	std::vector <DOFPointer> s_2_dof_list;
	s_2_dof_list.push_back(s_center_);
	s_2_dof_list.push_back(radius_);
	s_2_dof_list.push_back(theta_2_);

	std::vector <DOFPointer> t_2_dof_list;
	t_2_dof_list.push_back(t_center_);
	t_2_dof_list.push_back(radius_);
	t_2_dof_list.push_back(theta_2_);

	// create dependent DOF's based on the above expressions
	DOFPointer s_dof(new DependentDOF(s_2, s_2_dof_list));
	DOFPointer t_dof(new DependentDOF(t_2, t_2_dof_list));

	// create the actual point object
	Point2DPointer result(new Point2D(s_dof, t_dof, sketch_plane_));
	return result;
}

void Arc2D::GetTangent1(GiNaC::ex & s_component, GiNaC::ex & t_component)
{
	s_component = sin(GetTheta1()->GetVariable());
	t_component = -cos(GetTheta1()->GetVariable());
}

void Arc2D::GetTangent2(GiNaC::ex & s_component, GiNaC::ex & t_component)
{
	s_component = -sin(GetTheta2()->GetVariable());
	t_component = cos(GetTheta2()->GetVariable());
}

File: trunk/src/mono/ConstraintSolver/Arc2D.cs

using System;
using pSketcher.ConstraintSolver.SolverFunctions;

namespace pSketcher.ConstraintSolver
{

    /// <summary>
    /// 2D Arc Primitive
    /// </summary>
    class Arc2D : Edge2D
    {
        //
        // Field Data
        //
        Point2D point1_;
        Point2D point2_;
        Vector2D tangent1_;
        Vector2D tangent2_;
        
        //
        // Constructors
        //
        public Arc2D (SketchPlane sketch_plane, double s_center, double t_center, 
                     double theta1, double theta2, double radius, 
                     bool s_center_free, bool t_center_free, bool theta1_free,
                     bool theta2_free, bool radius_free) : 
            base (sketch_plane)
        {
            CenterPoint = new Point2D(SketchPlane, s_center, t_center, s_center_free, t_center_free);
            Theta1 = new IndependentDOF(theta1, theta1_free);
            Theta2 = new IndependentDOF(theta2, theta2_free);
            Radius = new IndependentDOF(radius, radius_free);
            
            // define the end points
            // end point 1
            DependentDOF s_dof = new DependentDOF(new arc2d_point_s(CenterPoint.S,Radius,Theta1));
            DependentDOF t_dof = new DependentDOF(new arc2d_point_t(CenterPoint.T,Radius,Theta1));            
            
            point1_ = new Point2D(SketchPlane, s_dof, t_dof);
            
            // end point 2
            s_dof = new DependentDOF(new arc2d_point_s(CenterPoint.S,Radius,Theta2));
            t_dof = new DependentDOF(new arc2d_point_t(CenterPoint.T,Radius,Theta2));            
            
            point2_ = new Point2D(SketchPlane, s_dof, t_dof);
            
            // define the tangent vectors
            // tangent vector for end point 1
            s_dof = new DependentDOF(new arc2d_tangent_s(Theta1));
            t_dof = new DependentDOF(new arc2d_tangent_t(Theta1));
        
            tangent1_ = new Vector2D(sketch_plane,s_dof,t_dof);   
            
            // tangent vector for end point 2
            s_dof = new DependentDOF(new arc2d_tangent_s(Theta2));
            t_dof = new DependentDOF(new arc2d_tangent_t(Theta2));
        
            tangent2_ = new Vector2D(sketch_plane,s_dof,t_dof);
        }
        
        //
        // Properties and Accessors
        //
        public override Point2D Point1 { get {return point1_;} }
        public override Point2D Point2 { get {return point2_;} }
        public override Vector2D Tangent1 { get {return tangent1_;} }
        public override Vector2D Tangent2 { get {return tangent2_;} }
        
        public Point2D CenterPoint { get; private set; }
        public DOF Radius { get; private set;}
        public DOF Theta1 { get; private set; }
        public DOF Theta2 { get; private set; }
        
    } // Class Arc2D
}
